In [ ]:
import pandas as pd

# Veri setini bir Excel dosyasından yükleyelim
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/00352/Online%20Retail.xlsx'
df = pd.read_excel(url)

# Veri setinin ilk 5 satırı
print(df.head())
print("")

# Sütunların veri tipleri
print(df.dtypes)
print("")

# veri setinin boyutu
print(df.info())
print("")

# Her sütundaki eksik değerlerin sayısı
print(df.isnull().sum())



  InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  

InvoiceNo              object
StockCode              object
Description            object
Quantity                int64
InvoiceDate    datetime64[ns]
UnitPrice             float64
Customer

In [ ]:
import pandas as pd

# Colab'daki dosya yolu
file_path = "Online Retail.xlsx"

# Veri setini yükleme
df = pd.read_excel(file_path)

# CustomerID sütununda eksik değerlerin silinmesi
df_cleaned = df.dropna(subset=['CustomerID'])

# Güncellenmiş veri setini kaydetme
df_cleaned.to_excel("1-) Cleaned Data Set.xlsx", index=False)

print("Güncellenmiş veri seti 'Cleaned Data Set.xlsx' olarak kaydedildi.")


Güncellenmiş veri seti 'Cleaned Data Set.xlsx' olarak kaydedildi.


In [ ]:
import pandas as pd

# Güncellenmiş veri seti
file_path = "1-) Cleaned Data Set.xlsx"
df_cleaned = pd.read_excel(file_path)

# CustomerID sütununda eksik değerlerin silinmesi
df_cleaned = df.dropna(subset=['Description'])

# CustomerID ve Description sütunlarında eksik değerler silindi.
df_cleaned.to_excel("2-)Description and CustomerID Filled.xlsx", index=False)


print("Güncellenmiş veri seti '2-)Description and CustomerID Filled.xlsx' olarak kaydedildi.")


Güncellenmiş veri seti '2-)Description and CustomerID Filled.xlsx' olarak kaydedildi.


In [ ]:
import pandas as pd

file_path = "1-) Cleaned Data Set.xlsx"
df = pd.read_excel(file_path)

# Güncel veri setinin boyutu
print("Veri setinin boyutu (satır, sütun):", df.shape)
print(df.info())

# Güncel veri setinde eksik değer var mı yok mu tekrar kontrol ediyoruz
print(df.isnull().sum())


Veri setinin boyutu (satır, sütun): (406829, 8)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 406829 entries, 0 to 406828
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    406829 non-null  object        
 1   StockCode    406829 non-null  object        
 2   Description  406829 non-null  object        
 3   Quantity     406829 non-null  int64         
 4   InvoiceDate  406829 non-null  datetime64[ns]
 5   UnitPrice    406829 non-null  float64       
 6   CustomerID   406829 non-null  int64         
 7   Country      406829 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(2), object(4)
memory usage: 24.8+ MB
None
InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64


In [ ]:
import pandas as pd

file_path = "1-) Cleaned Data Set.xlsx"
df = pd.read_excel(file_path)

# Yinelenen satırları kontrol etme (tamamıyla aynı olanalra bakıyorum)
duplicate_count = df.duplicated().sum()
print("Tamamen aynı olan yinelenen satır sayısı:", duplicate_count)

# Eğer tamamen aynı satırlar ise
if duplicate_count > 0:
    duplicates = df[df.duplicated()]
    print("Tamamen aynı olan yinelenen satırlar:")
    print(duplicates)

# Tamamen aynı olan satırları silme
df = df.drop_duplicates()
print("Tamamen aynı olan yinelenen satırlar silindi. Güncel veri seti boyutu:", df.shape)

# Temizlenmiş veri seti
df.to_excel("3-)No Duplicates.xlsx", index=False)

print("Güncellenmiş yeni veri seti '3-)No Duplicates.xlsx' olarak kaydedildi.")


Tamamen aynı olan yinelenen satır sayısı: 5225
Tamamen aynı olan yinelenen satırlar:
       InvoiceNo StockCode                        Description  Quantity  \
517       536409     21866        UNION JACK FLAG LUGGAGE TAG         1   
527       536409     22866      HAND WARMER SCOTTY DOG DESIGN         1   
537       536409     22900    SET 2 TEA TOWELS I LOVE LONDON          1   
539       536409     22111       SCOTTIE DOG HOT WATER BOTTLE         1   
555       536412     22327  ROUND SNACK BOXES SET OF 4 SKULLS         1   
...          ...       ...                                ...       ...   
406595    581538     22068        BLACK PIRATE TREASURE CHEST         1   
406609    581538     23318     BOX OF 6 MINI VINTAGE CRACKERS         1   
406612    581538     22992             REVOLVER WOODEN RULER          1   
406619    581538     22694                       WICKER STAR          1   
406621    581538     23343       JUMBO BAG VINTAGE CHRISTMAS          1   

              

# Negatif Değerlerin Anlamı ve Kontrol Edilmesi
1. Quantity (Miktar):
  * Bu sütun, işlemlerde alınan ürün miktarını ifade eder.
  * Negatif değerler, ürün iadesini veya yanlış bir veri girişini temsil edebilir. Örneğin, müşteri ürünü iade ettiğinde, satın alma miktarı negatif olarak kaydedilebilir.
  * Eğer analiz sadece satışlara odaklanıyorsa, bu negatif değerler çıkarılmalıdır. Ancak, analiziniz iade işlemlerini de kapsıyorsa, bu veriler tutulabilir.
2. UnitPrice (Birim Fiyat):
  * Bu sütun, bir ürünün birim başına fiyatını ifade eder.
  * Negatif değerler, veri girişinde bir hata veya yanlış bir işlem olduğunu gösterir. Ürünlerin fiyatlarının negatif olması mantıklı değildir.
  * Bu nedenle, UnitPrice sütunundaki negatif değerler her durumda temizlenmelidir.

In [ ]:
import pandas as pd

file_path = "1-) Cleaned Data Set.xlsx"
df = pd.read_excel(file_path)

# UnitPrice'da negatif değerler
negative_unitprice = df[df['UnitPrice'] < 0]
print("Negatif UnitPrice satır sayısı:", len(negative_unitprice))
# Temizlenmiş veri setinin boyutu

print("Güncel veri seti boyutu:", df.shape)

Negatif UnitPrice satır sayısı: 0
Güncel veri seti boyutu: (406829, 8)


In [ ]:
import pandas as pd

file_path = "1-) Cleaned Data Set.xlsx"
df = pd.read_excel(file_path)

# Quantity'de negatif değerleri kontrol etme
negative_quantity = df[df['Quantity'] < 0]
print("Negatif Quantity satır sayısı:", len(negative_quantity))



Negatif Quantity satır sayısı: 8905


In [ ]:
import pandas as pd

file_path = "1-) Cleaned Data Set.xlsx"
df = pd.read_excel(file_path)

# Quantity'de negatif değerleri kontrol etme
negative_quantity = df[df['Quantity'] < 0]
print("Negatif Quantity satır sayısı:", len(negative_quantity))

# Eğer negatif Quantity değerleri var ise
if len(negative_quantity) > 0:
    print("Negatif Quantity değerleri:")
    print(negative_quantity)

# Negatif Quantity değerlerini çıkarma
df_cleaned = df[df['Quantity'] >= 0]

# Güncellenmiş veri setinin boyutunu kontrol et
print("Güncel veri seti boyutu:", df_cleaned.shape)

# Güncellenmiş veri setini kaydetme
cleaned_file_path = "4-)Quantity Values Positive.xlsx"
df_cleaned.to_excel(cleaned_file_path, index=False)

print("Güncel veri seti '4-)Quantity Values Positive.xlsx' olarak kaydedildi.")


Negatif Quantity satır sayısı: 8905
Negatif Quantity değerleri:
       InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
406377   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
406461   C581499         M                            Manual        -1   
406635   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
406636   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
406637   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

               InvoiceDate  UnitPrice  Customer

In [ ]:
import pandas as pd

file_path = "4-)Quantity Values Positive.xlsx"
df = pd.read_excel(file_path)

# Yeni özellik: TotalPrice (Toplam Fiyat) --> TotalPrice=Quantity*UnitPrice
df['TotalPrice'] = df['Quantity'] * df['UnitPrice']

# Güncellenmiş veri setini kaydetme
df.to_excel("5-)Feature Engineered.xlsx", index=False)

print("Güncellenmiş veri seti '5-)Feature Engineered.xlsx' olarak kaydedildi.")


Güncellenmiş veri seti '5-)Feature Engineered.xlsx' olarak kaydedildi.


In [ ]:
import pandas as pd

file_path = "5-)Feature Engineered.xlsx"
df = pd.read_excel(file_path)

# Veri setinin  satır ve sütun sayısı
print(f"Toplam satır sayısı: {df.shape[0]}")
print(f"Toplam sütun sayısı: {df.shape[1]}")

# Sütun adlarını ve veri türleri
print("\nSütun Adları ve Veri Türleri:")
print(df.dtypes)
print("")
print(df.head())


Toplam satır sayısı: 397924
Toplam sütun sayısı: 9

Sütun Adları ve Veri Türleri:
InvoiceNo               int64
StockCode              object
Description            object
Quantity                int64
InvoiceDate    datetime64[ns]
UnitPrice             float64
CustomerID              int64
Country                object
TotalPrice            float64
dtype: object

   InvoiceNo StockCode                          Description  Quantity  \
0     536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1     536365     71053                  WHITE METAL LANTERN         6   
2     536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3     536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4     536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  TotalPrice  
0 2010-12-01 08:26:00       2.55       17850  United Kingdom       15.30  
1 2010-12-01 08:26:00       3.39       17850

*  Veri setinde normalizasyona ihtiyaç duyulmasının temel nedenleri, veri analizi, makine öğrenimi modelleri ve istatistiksel yöntemlerin daha verimli ve doğru bir şekilde çalışmasını sağlamaktır.

1. Özelliklerin Farklı Ölçeklerde Olması
Veri setindeki sütunlar (özellikler) genellikle farklı ölçüm birimlerine veya aralıklara sahip olabilir.
Örneğin:
Ağırlık (kg cinsinden, 0-100 arasında olabilir).
Maaş (₺ cinsinden, 1000-100000 arasında olabilir).
Eğer bu özellikler aynı ölçeğe getirilmezse, bazı algoritmalar büyük değerlere sahip sütunları daha fazla önemser, bu da hatalı sonuçlara yol açabilir.
Örnek: Bir sınıflandırma algoritması, "Maaş" sütununu "Ağırlık" sütunundan daha önemli gibi görebilir sadece maaş değerleri daha büyük olduğu için.

2. Algoritmaların Daha İyi Çalışması
Mesafe tabanlı algoritmalar (örn. K-En Yakın Komşu, K-Means, Destek Vektör Makineleri), iki veri noktası arasındaki mesafeyi hesaplamak için özelliklerin ölçeğine duyarlıdır.
Eğer veri setinde bir özellik diğerlerinden çok daha büyük ölçeklere sahipse, bu algoritma o özelliği daha baskın hale getirir.
Bu yüzden, tüm özellikleri benzer bir aralığa (örneğin 0-1) dönüştürmek gereklidir.
Örnek: K-Means algoritması, kümeleri belirlerken özelliklerin ölçeğine çok duyarlıdır. Normalizasyon yapılmadığında yanlış kümeler oluşturabilir.

3. Daha Hızlı ve Stabil Model Eğitimini Sağlamak
Makine öğrenimi modelleri, özellikle gradyan tabanlı optimizasyon kullanan algoritmalar (örneğin Yapay Sinir Ağları, Lojistik Regresyon) için, normalizasyon hesaplamaların daha hızlı ve stabil olmasını sağlar.
Büyük değerler, modelin ağırlık güncellemelerini bozabilir veya gradyan patlamasına yol açabilir.
Küçük ve dengeli değerlerle model daha hızlı öğrenir.
4. Aykırı Değerlerin Etkisini Azaltmak
Özellikle normalizasyon ile birlikte kullanılan standartlaştırma (z-score normalization) yöntemleri, veri setindeki aykırı değerlerin etkisini azaltabilir. Bu, özelliklerin ortalamasını ve standart sapmasını hesaba katarak veriyi ölçeklendirir.
5. Karşılaştırılabilirlik Sağlamak
Farklı ölçeklerdeki özellikleri karşılaştırmak zordur. Normalizasyon, tüm özellikleri aynı ölçeğe çekerek bu karşılaştırılabilirliği sağlar.
Örnek: Bir veri setindeki "Gelir" ve "Yaş" sütunlarını anlamlı bir şekilde karşılaştırmak için normalizasyon gerekebilir.

6. Model Performansını Artırmak
Normalizasyon, bazı modellerin performansını artırabilir çünkü özellikler arasındaki farkların daha iyi anlaşılmasını sağlar.
Özellikle derin öğrenme modelleri ve regresyon modelleri, normalizasyonla daha doğru sonuçlar verir.
7. Çıktının Yorumlanabilirliğini Artırmak
Normalizasyon, model çıktılarını yorumlamayı kolaylaştırabilir. Özellikle 0-1 aralığına getirilen veriler, anlam açısından daha tutarlı hale gelir.


**Normalizasyon Çeşitleri**
1. Min-Max Normalizasyonu: Veriyi 0 ile 1 arasında ölçekler.
  * x'= (x-min(x))/(max(x)-min(x))

2. Z-Score Normalizasyonu: Veriyi standart normal dağılıma uyarlar (μ=0,σ=1)
  * x' = (x-μ)/σ

3. Decimal Scaling: Veriyi 10'un kuvvetleriyle ölçekler.

In [ ]:
import pandas as pd

file_path = "5-)Feature Engineered.xlsx"
df = pd.read_excel(file_path)

# Normalizasyon uygulanacak sütunlar
columns_to_normalize = ['Quantity', 'UnitPrice', 'TotalPrice']

# Her sütun için Min ve Max değerleri
min_max_values = {}
for column in columns_to_normalize:
    min_val = df[column].min()
    max_val = df[column].max()
    min_max_values[column] = {'min': min_val, 'max': max_val}


for column in columns_to_normalize:
    min_val = min_max_values[column]['min']
    max_val = min_max_values[column]['max']
    df[column] = (df[column] - min_val) / (max_val - min_val)

# Normalizasyon sonrası veri setini kaydetme işlemi
normalized_file_path = "6-)Data Set Normalized.xlsx"
df.to_excel(normalized_file_path, index=False)


print("Normalizasyon tamamlandı, veri seti '6-)Data Set Normalized.xlsx' olarak kaydedildi.")
print("Min ve Max değerleri (Normalizasyon öncesi):")
for column, values in min_max_values.items():
    print(f"{column}: Min = {values['min']}, Max = {values['max']}")


Normalizasyon tamamlandı, veri seti '6-)Data Set Normalized.xlsx' olarak kaydedildi.
Min ve Max değerleri (Normalizasyon öncesi):
Quantity: Min = 1, Max = 80995
UnitPrice: Min = 0.0, Max = 8142.75
TotalPrice: Min = 0.0, Max = 168469.6


In [ ]:
import pandas as pd
from sklearn.preprocessing import MinMaxScaler

file_path = "5-)Feature Engineered.xlsx"
df = pd.read_excel(file_path)

# normalize edilen sutünlar
columns_to_normalize = ['Quantity', 'UnitPrice', 'TotalPrice']


scaler = MinMaxScaler()

# yeni sutünlar olşturrmak
for column in columns_to_normalize:
    normalized_column_name = f"{column}Normalized"
    df[normalized_column_name] = scaler.fit_transform(df[[column]])


normalized_file_path = "7-)Feature Engineered with Normalized Columns.xlsx"
df.to_excel(normalized_file_path, index=False)

print(f"Yeni veri seti '7-)Feature Engineered with Normalized Columns.xlsx' olarak kaydedildi.")


Yeni veri seti '7-)Feature Engineered with Normalized Columns.xlsx' olarak kaydedildi.


Bu hesaplamaları yapmanızın ve bu büyük veri seti üzerinde çalışmanızın nedenleri aşağıda açıklanmıştır:

1. Veri Dağılımını Anlamak:
Ortalama (mean), medyan (median), standart sapma (std), minimum (min) ve maksimum (max) değerleri, veri setindeki sütunların dağılımını anlamak için kullanılır. Bu, verilerin genel özelliklerini özetler ve normalizasyonun etkisini değerlendirmenizi sağlar.
2. Aykırı Değerleri Tespit Etmek:
Standart sapma ve maksimum değerler, veri setinde aykırı değerlerin (outliers) olup olmadığını anlamak için önemlidir. Örneğin, "Quantity" sütunundaki 80,995 gibi yüksek bir değer, veri setinde bir aykırılık olduğunu gösterebilir.
3. Normalizasyonun Doğruluğunu Kontrol Etmek:
Normalizasyon öncesi değerlerle karşılaştırma yaparak, verilerin doğru ölçeklendiğinden emin olabilirsiniz. Minimum ve maksimum değerlerin 0 ile 1 arasında olması beklenir.
4. Büyük Veri Setlerinde İstatistiksel Özetleme:
392,738 satırlık bir veri setiyle çalışırken, tüm satırları analiz etmek yerine bu özet istatistiklerle verinin genel eğilimlerini daha hızlı görebiliriz.
5. Model Seçimi ve Hazırlık:
Makine öğrenimi veya veri analizi modelleri için, özelliklerin ölçeğini ve dağılımını anlamak, modelin daha iyi performans göstermesi için önemlidir.
6. Karar Verme:
Örneğin, "UnitPrice" sütununda standart sapma yüksekse, fiyatlardaki çeşitliliğin fazla olduğunu anlar ve farklı stratejiler uygulayabilirsiniz (örneğin, veriyi gruplayabilirsiniz).
7. Sonuç:
Bu tür hesaplamalar, büyük bir veri setinde normalizasyon sonrası süreci değerlendirmenize, verinin genel yapısını anlamanıza ve analitik ya da modelleme kararlarını daha bilinçli bir şekilde vermenize olanak tanır.

In [ ]:
import pandas as pd

file_path = "7-)Feature Engineered with Normalized Columns.xlsx"
df = pd.read_excel(file_path)

# Normalizasyon öncesi sütunlar
columns_original = ['Quantity', 'UnitPrice', 'TotalPrice']

# Normalizasyon sonrası sütunlar
columns_normalized = ['QuantityNormalized', 'UnitPriceNormalized', 'TotalPriceNormalized']

# Hesaplamalar: Normalizasyon öncesi
summary_stats_original = df[columns_original].agg(['mean', 'median', 'std', 'min', 'max'])

# Hesaplamalar: Normalizasyon sonrası
summary_stats_normalized = df[columns_normalized].agg(['mean', 'median', 'std', 'min', 'max'])


print("Normalizasyon öncesi sütunların hesaplamaları:")
print(summary_stats_original)
print("")

print("\nNormalizasyon sonrası sütunların hesaplamaları:")
print(summary_stats_normalized)


Normalizasyon öncesi sütunların hesaplamaları:
            Quantity    UnitPrice     TotalPrice
mean       13.021823     3.116174      22.394749
median      6.000000     1.950000      11.800000
std       180.420210    22.096788     309.055588
min         1.000000     0.000000       0.000000
max     80995.000000  8142.750000  168469.600000


Normalizasyon sonrası sütunların hesaplamaları:
        QuantityNormalized  UnitPriceNormalized  TotalPriceNormalized
mean              0.000148             0.000383              0.000133
median            0.000062             0.000239              0.000070
std               0.002228             0.002714              0.001834
min               0.000000             0.000000              0.000000
max               1.000000             1.000000              1.000000


In [ ]:
import pandas as pd

file_path = "7-)Feature Engineered with Normalized Columns.xlsx"
df = pd.read_excel(file_path)

# Benzersiz ülkeler ve 3 basamaklı kodlar oluştur
unique_countries = pd.DataFrame({
    'Country': df['Country'].unique(),
    'Code': [format(i + 1, '03d') for i in range(len(df['Country'].unique()))]
})

# Ülke adlarını 3 basamaklı kodlara dönüştürmek için bir eşleme haritası oluştur
country_to_numeric_map = dict(zip(unique_countries['Country'], unique_countries['Code']))

# 'Country' sütununu 3 basamaklı kodlarla güncelle
df['Country'] = df['Country'].map(country_to_numeric_map)

# Güncellenmiş dosyayı kaydet
output_file_path = "8-)Country_with_Numeric_Value_Updated.xlsx"
df.to_excel(output_file_path, index=False)

# Ülkeler ve kodlar tablosunu göster
print("Ülke ve Kod Tablosu (001, 002, ...):")
print(unique_countries)


Ülke ve Kod Tablosu (001, 002, ...):
                 Country Code
0         United Kingdom  001
1                 France  002
2              Australia  003
3            Netherlands  004
4                Germany  005
5                 Norway  006
6                   EIRE  007
7            Switzerland  008
8                  Spain  009
9                 Poland  010
10              Portugal  011
11                 Italy  012
12               Belgium  013
13             Lithuania  014
14                 Japan  015
15               Iceland  016
16       Channel Islands  017
17               Denmark  018
18                Cyprus  019
19                Sweden  020
20               Finland  021
21               Austria  022
22                Greece  023
23             Singapore  024
24               Lebanon  025
25  United Arab Emirates  026
26                Israel  027
27          Saudi Arabia  028
28        Czech Republic  029
29                Canada  030
30           Unspecified  031
31 